gwcs/wcs.py

```diff
@@ -332,7 +388,502 @@ def _invert(self, *args, **kwargs):
+            while (ind.shape[0] > 0 and k < maxiter):
+                # Find correction to the previous solution:
+                dpixnew = correction(pix[ind])
+
+                # Compute norm (L2) of the correction:
+                dnnew = np.sum(np.square(dpixnew), axis=1)
+
+                # Bookeeping of corrections:
+                dnprev[ind] = dn[ind].copy()
+                dn[ind] = dnnew
+
+                if detect_divergence:
+                    # Find indices of pixels that are converging:
+                    conv = np.logical_or(dnnew < dnprev[ind], dnnew < tol2)
+                    if not np.all(conv):
+                        conv = np.ones_like(dnnew, dtype=np.bool)
+                        # raise ValueError("vvvvvvvvvv")
+                    iconv = np.where(conv)
+                    iiconv = ind[iconv]
+
+                    # Apply correction:
+                    dpixgood = dpixnew[iconv]
+                    pix[iiconv] -= dpixgood
+                    dpix[iiconv] = dpixgood
+
+                    # Find indices of solutions that have not yet
+                    # converged to the requested accuracy
+                    # AND that do not diverge:
+                    subind, = np.where((dnnew >= tol2) & conv)
+
+                else:
+                    # Apply correction:
+                    pix[ind] -= dpixnew
+                    dpix[ind] = dpixnew
+
+                    # Find indices of solutions that have not yet
+                    # converged to the requested accuracy:
+                    subind, = np.where(dnnew >= tol2)
+
+                # Choose solutions that need more iterations:
+                ind = ind[subind]
+                world = world[subind]
+
+                k += 1
+
+        # ############################################################
+        # #         FINAL DETECTION OF INVALID, DIVERGING,          ##
+        # #         AND FAILED-TO-CONVERGE POINTS                   ##
+        # ############################################################
+        # Identify diverging and/or invalid points:
+        invalid = ((~np.all(np.isfinite(pix), axis=1)) &
+                   (np.all(np.isfinite(world0), axis=1)))
+
+        # When detect_divergence==False, dnprev is outdated
+        # (it is the norm of the very first correction).
+        # Still better than nothing...
+        inddiv, = np.where(((dn >= tol2) & (dn >= dnprev)) | invalid)
+        if inddiv.shape[0] == 0:
+            inddiv = None
+
+        # Identify points that did not converge within 'maxiter'
+        # iterations:
+        if k >= maxiter:
+            ind, = np.where((dn >= tol2) & (dn < dnprev) & (~invalid))
+            if ind.shape[0] == 0:
+                ind = None
+        else:
+            ind = None
+
+        # Restore previous numpy error settings:
+        np.seterr(invalid=old_invalid, over=old_over)
+
+        # ############################################################
```

Comment:
Is there an option to get the results with a flag of points that didn't converge? I'm thinking of a case where just a few of many don't converge and now the user is left to puzzle which ones were the problems.

Reply:
No there is no such option. If you turn on `detect_divergence` (which is kind of natural to do if you want to know which points diverged) and set `quiet=False`, the code will raise an exception `NoConvergence` that you can catch and examine it. The `divergent` attribute of the exception will give you that info:

```
    divergent : None, `numpy.ndarray`
        Indices of the points in ``best_solution`` array
        for which the solution appears to be divergent. If the
        solution does not diverge, ``divergent`` will be set to `None`.
```

Reply:
Good. But where will you document this?  It is only mentioned in the exception class as far as I can see and I'm not sure most people will look there, and it should be made more transparent that they can use this to find what points are the problem.

Reply:
As mentioned before (and off-line) at this stage I was more interested in getting the concept and not spend time on documenting things that may change.

This vectorized approach being essentially identical to the one in `astropy.wcs.WCS.all_world2pix` please take a look at https://docs.astropy.org/en/stable/api/astropy.wcs.WCS.html#astropy.wcs.WCS.all_world2pix

Reply:
Hm... It seems I was partially incorrect:

> Even when detect_divergence is False, all_world2pix(), at the end of the iterative process, will identify invalid results (NaN or Inf) as “diverging” solutions and will raise NoConvergence unless the quiet parameter is set to True.